stake-pool/birdbath/src/processor.rs

```diff
@@ -0,0 +1,1100 @@
+                pool_mint_info.key,
+                mint_rent,
+                mint_space as u64,
+                token_program_info.key,
+            ),
+            &[
+                payer_info.clone(),
+                pool_mint_info.clone(),
+                system_program_info.clone(),
+            ],
+            mint_signers,
+        )?;
+
+        invoke_signed(
+            &spl_token::instruction::initialize_mint(
+                token_program_info.key,
+                pool_mint_info.key,
+                pool_authority_info.key,
+                None,
+                MINT_DECIMALS,
+            )?,
+            &[
+                pool_mint_info.clone(),
+                rent_info.clone(),
+                system_program_info.clone(),
+            ],
+            authority_signers,
+        )?;
+
+        // create the pool stake account
+        let stake_space = std::mem::size_of::<stake::state::StakeState>();
+        let required_lamports = rent.minimum_balance(stake_space).saturating_add(1);
+        let authorized = stake::state::Authorized::auto(pool_authority_info.key);
+
+        invoke_signed(
+            &system_instruction::create_account(
+                payer_info.key,
+                pool_stake_info.key,
+                required_lamports,
+                stake_space as u64,
+                stake_program_info.key,
+            ),
+            &[
+                payer_info.clone(),
+                pool_stake_info.clone(),
+                stake_program_info.clone(),
+            ],
+            stake_signers,
+        )?;
+
+        invoke_signed(
+            &stake::instruction::initialize_checked(pool_stake_info.key, &authorized),
+            &[
+                pool_stake_info.clone(),
+                rent_info.clone(),
+                pool_authority_info.clone(),
+                pool_authority_info.clone(),
+            ],
+            authority_signers,
+        )?;
+
+        // delegate the stake so it activates
+        invoke_signed(
+            &stake::instruction::delegate_stake(
+                pool_stake_info.key,
+                pool_authority_info.key,
+                validator_vote_info.key,
+            ),
+            &[
+                pool_stake_info.clone(),
+                validator_vote_info.clone(),
+                clock_info.clone(),
+                stake_history_info.clone(),
+                stake_config_info.clone(),
+                pool_authority_info.clone(),
+            ],
+            authority_signers,
+        )?;
+
+        // could mint the token here if we wanted, either to user or incinerator
+
+        Ok(())
+    }
+
+    // XXX ok cool next up ummm
+    // the other two functions are extremely simplified version of their namesakes
+    // for deposit we literally only need to call stake_merge (or a hana version), not authorize
+    // because the user can et both authorities to ours rather than going through deposit authority
```

Comment:
This does simplify things a lot! And since you check the address of the destination, I think you're totally safe. We'll need to make sure it doesn't open up any attacks from people trying to merge accounts owned by the pool, but I don't think it's possible

Reply:
yea theres definitely test cases id like to cover of users trying to do tricky shit but... im pretty sure this is all safe, because ultimately every vote account gets one and only one canonical stake account, and every stake account has a different authority